### PR TITLE
batches: support `replaceBatchSpecInput` with existing draft batch change

### PR DESCRIPTION
**Pull request by courier-new**
Supersedes #28827 (and to a lesser extent, #28825, until we resolve what to do for executing a new batch spec on a non-draft batch change).

Updates `replaceBatchSpecInput` to re-associate the previous batch spec's batch change with the new one that is replacing it, but only if the batch change was still a draft. It will produce an error if you attempt to call it with a batch spec that was already applied to a batch change.

To do so, I stole `BatchChange.IsDraft` that Thorsten wrote from #28840. 😈

### CODE REVIEW

File: enterprise/internal/batches/types/batch_change.go [hunk 1]
```diff
@@ -40,3 +40,13 @@ func (c *BatchChange) Clone() *BatchChange {
...
 func (c *BatchChange) Closed() bool { return !c.ClosedAt.IsZero() }
+
+// IsDraft returns true when the BatchChange is a draft ("shallow") Batch
+// Change, i.e. it's associated with a BatchSpec but it hasn't been applied
+// yet.
+func (c *BatchChange) IsDraft() bool {
+	return c.BatchSpecID != 0 &&
+		c.LastAppliedAt.IsZero() &&
+		c.LastApplierID == 0 &&
+		c.InitialApplierID == 0
```
Comment (LawnGnome) [inline enterprise/internal/batches/types/batch_change.go:51]: I don't feel at all strongly about this, but are these checks somewhat redundant? Could we just check `InitialApplierID` and be done with it?

**Review by LawnGnome — APPROVED** [review]
LGTM.

File: enterprise/internal/batches/types/batch_change.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/batches/types/batch_change.go:51]: Yeah, we talked about this on Monday. Fix was made in another PR here: https://github.com/sourcegraph/sourcegraph/pull/28971

Comment (LawnGnome) [reply enterprise/internal/batches/types/batch_change.go:51]: @courier-new gave me the heads up via DM after I made this review. All good!

**Comment by courier-new** [comment]
Superceded by #28829.

**Pull request closed without merging**